src/topology/scheme.lean

```diff
@@ -0,0 +1,252 @@
+/-
+Copyright (c) 2023 Felix Weilacher. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Felix Weilacher
+-/
+import topology.metric_space.pi_nat
+
+/-!
+# (Topological) Schemes and their induced maps
+
+In topology, and especially descriptive set theory, one often constructs functions `(ℕ → β) → α`,
+where α is some topological space and β is a discrete space, as an appropriate limit of some map
+`list β → set α`. We call the latter type of map a "`β`-scheme on `α`".
+
+This file develops the basic, abstract theory of these schemes and the functions they induce.
+
+## Main Definitions
+
+* `induced_map A` : The aforementioned "limit" of a scheme `A : list β → set α`.
+  This is a partial function from `ℕ → β` to `a`,
+  implemented here as an object of type `Σ s : set (ℕ → β), s → α`.
+  That is, `(induced_map A).1` is the domain and `(induced_map A).2` is the function.
+
+## Implementation Notes
+
+We consider end-appending to be the fundamental way to build lists (say on `β`) inductively,
+as this interacts better with the topology on `ℕ → β`.
+As a result, functions like `list.nth` or `stream.take` do not have their intended meaning
+in this file. We define an analogue of `stream.take`, `res`, for our purposes.
+It is related by the equation `res x n = (stream.take n x).reverse`
+
+## References
+
+* [kechris1995] (Chapters 6-7)
+
+## Tags
+
+Scheme.
+
+-/
+
+namespace scheme
```

Comment:
@leanprover-community/mathlib-ag, what do you think we should do to avoid ambiguity with your schemes? Should we put all that in the `topology` namespace?

Reply:
Can we replace `scheme` with something else?

Reply:
I think `topology.scheme` or `topology.metric_space.scheme` would be fine

Reply:
I don't. I think it would cause confusion for newcomers. I did a search for "scheme topology" and of course the only thing that came up was schemes in algebraic geometry. How common is this word in descriptive set theory?

Reply:
It is not that common. People usually say "Cantor Scheme" or "Lusin/Souslin scheme" for when beta is 2 or `nat` respectively.

Reply:
Similar objects have been called "approximations" before. I could base the name off that instead.

Reply:
I'll go with cantor_scheme for now...